### PR TITLE
add required field to parameters

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by liggitt** [comment]
@spadgett @jwforres FYI

**Comment by liggitt** [comment]
Process would fail if no value or generator was provided for a required value, right?

**Comment by bparees** [comment]
@jwforres @jcantrill fyi.

Note that *nothing* consumes these fields currently.

**Comment by bparees** [comment]
@liggitt that's the idea, someday.  for now this was requested by the eclipse team so they may be the first to actually respect the field.

**Comment by bparees** [comment]
@liggitt I discussed w/ @smarterclayton and he did not threaten my life when i suggested introducing the field w/o updating any clients to leverage the field. so.....

**Comment by liggitt** [comment]
Having process honor it first would at least prevent all other clients from creating objects from templates with parameter values contrary to the author's intent. Gradual roll-out between UI and Eclipse clients is fine

**Comment by liggitt** [comment]
It doesn't bother me to do the process bits in a follow-up PR, I'd just like to see that sooner than later.

**Comment by bparees** [comment]
@liggitt the template process endpoint will now return an error if a required parameter does not have a non-empty value. 

@mfojtik ptal.

File: pkg/template/template.go [hunk 1]
```diff
@@ -169,6 +169,10 @@ func (p *Processor) GenerateParameterValues(t *api.Template) error {
 				return fmt.Errorf("template.parameters[%v]: Unable to convert the generated value '%#v' to string", i, value)
 			}
 		}
+		if len(param.Value) == 0 && param.Required {
+			return fmt.Errorf("template.parameters[%v]: parameter is marked required but has no default value and no generator", i)
```
Comment (smarterclayton) [inline pkg/template/template.go]: Needs to be a structured error for the exact parameter the user set (have a sub field error indicated)

Comment (liggitt) [inline pkg/template/template.go]: That's probably true of all the parameter errors if we expect a client to be able to programmatically identify and present the error associated with the field

Comment (smarterclayton) [inline pkg/template/template.go]: Yeah.  At least fix this new one and have a follow up to fix the rest

On Jul 30, 2015, at 8:24 PM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/template/template.go
<https://github.com/openshift/origin/pull/3963#discussion_r35936756>:

> @@ -169,6 +169,10 @@ func (p *Processor) GenerateParameterValues(t *api.Template) error {
>  				return fmt.Errorf("template.parameters[%v]: Unable to convert the generated value '%#v' to string", i, value)
>  			}
>  		}
> +		if len(param.Value) == 0 && param.Required {
> +			return fmt.Errorf("template.parameters[%v]: parameter is marked required but has no default value and no generator", i)

That's probably true of all the parameter errors if we expect a client to
be able to programmatically identify and present the error associated with
the field

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3963/files#r35936756>.

Comment (bparees) [inline pkg/template/template.go]: do we have an example of this elsewhere?  I can create a new error struct but i'm not clear where that gets us in terms of manifesting it to the REST client.

Comment (liggitt) [inline pkg/template/template.go]: Is there a reason we can't use fielderrors.NewRequired, etc?

**Comment by bparees** [comment]
@liggitt @smarterclayton updated...  is that the type of error you wanted or were you looking for something more specific?

also note I added a tags field per the request of @jcantrill

**Comment by smarterclayton** [comment]
What is the use case for the tags field?

**Comment by bparees** [comment]
"(10:08:32 AM) jcantril1: they want to use it to be able to categorize parameters
(10:09:28 AM) jcantril1: so for instance if you wanted to present the user all the params that are related to 'database'"

**Comment by smarterclayton** [comment]
That is a really weak justification.  What is "database"?  Why would use
tags and not a formal group (like a fieldset in html)?  I would like to
have a use case to drive this that's pretty concrete and is the right
solution.  Adding things to template has a high bar.

On Jul 31, 2015, at 11:41 AM, Ben Parees <notifications@github.com> wrote:

"(10:08:32 AM) jcantril1: they want to use it to be able to categorize
parameters
(10:09:28 AM) jcantril1: so for instance if you wanted to present the user
all the params that are related to 'database'"

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3963#issuecomment-126727808>.

**Comment by liggitt** [comment]
Yeah, I don't like tags for that. At most, you'd want a single group for parameters (you would never show a parameter twice in a create form). I'd rather not rush into a fieldgroup concept without thinking through all the use cases.

**Comment by bparees** [comment]
ok i'm fine to remove it.  @jcantrill this is your fight if you want it. :)

**Comment by bparees** [comment]
> That's not the right format for a field name

@smarterclayton you want it to be [%d] w/ the index instead?  i figured the name was a little more useful when the error is rendered to a user, unless we assume the client tools are going to translate it.

**Comment by bparees** [comment]
based on this: https://github.com/openshift/origin/blob/master/pkg/image/api/validation/validation.go#L79

i'm going with:
			return fielderrors.NewFieldRequired(fmt.Sprintf("parameters[%s].value", param.Name))

**Comment by liggitt** [comment]
Can the name contain [ or ]?

**Comment by bparees** [comment]
no.

has to match:
var parameterNameExp = regexp.MustCompile(`^[a-zA-Z0-9\_]+$`)

**Comment by smarterclayton** [comment]
We use indexes to describe which parameter, not the name, since parameters
is an array in json.  You can add the name in the message if you want

On Jul 31, 2015, at 12:21 PM, Ben Parees <notifications@github.com> wrote:

based on this:
https://github.com/openshift/origin/blob/master/pkg/image/api/validation/validation.go#L79

i'm going with:
return fielderrors.NewFieldRequired(fmt.Sprintf("parameters[%s].value",
param.Name))

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3963#issuecomment-126739174>.

**Comment by liggitt** [comment]
The index is more correct… the ImageStream example is actually wrong. That said, the error is more usable in it's raw form when the name is part of the field name. Depends what we want to support long term, since this will be a likely place where we'll extract errors and correlate them with inputs to present to the user

**Comment by liggitt** [comment]
NewFieldRequired doesn't let you set a message

**Comment by smarterclayton** [comment]
Maybe it should.  However, we define a standard for this (machine readable
json path subset string) and so we should use the standard.  If we need
more strings to clarify for humans we should add those (or fix oc and the
web)

On Jul 31, 2015, at 12:26 PM, Jordan Liggitt <notifications@github.com>
wrote:

NewFieldRequired doesn't let you set a message

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3963#issuecomment-126741208>.

**Comment by bparees** [comment]
what about this
			err := fielderrors.NewFieldRequired(fmt.Sprintf("parameters[%d].value", i))
			err.Detail = fmt.Sprintf("parameter %s is required and has no default value and no generator or the generator produced an empty value")

**Comment by smarterclayton** [comment]
that's a helluva message.  Maybe "parameter %s is required and must be
specified"

On Fri, Jul 31, 2015 at 12:33 PM, Ben Parees <notifications@github.com>
wrote:

> what about this
> err := fielderrors.NewFieldRequired(fmt.Sprintf("parameters[%d].value", i))
> err.Detail = fmt.Sprintf("parameter %s is required and has no default
> value and no generator or the generator produced an empty value")
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3963#issuecomment-126746155>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by liggitt** [comment]
that seems too long if we would want a UI to show that to the user. "%s is a required parameter" seems ok to me

**Comment by liggitt** [comment]
Heh, ok… clearly either Clayton or I should stop commenting.

**Comment by bparees** [comment]
bah.  fine.

**Comment by bparees** [comment]
updated.

**Comment by smarterclayton** [comment]
One of us is on vacation.  The other two are people who want to be on
vacation.

On Fri, Jul 31, 2015 at 12:39 PM, Ben Parees <notifications@github.com>
wrote:

> bah. fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3963#issuecomment-126747377>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bparees** [comment]
everyone happy?  or at least not unhappy? :)

**Comment by liggitt** [comment]
Ecstatic

**Comment by bparees** [comment]
[test]

**Comment by jcantrill** [comment]
fighting for categories startes here https://github.com/openshift/origin/issues/3973

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
[merge]

**Pull request merged**